src/ServiceControl.Monitoring/EndpointInstanceId.cs

```diff
@@ -0,0 +1,69 @@
+﻿namespace ServiceControl.Monitoring
+{
+    using System.Collections.Generic;
+    using NServiceBus;
+
+    public class EndpointInstanceId
+    {
+        public string EndpointName { get; }
+        public string InstanceId { get; }       
+        public string InstanceName { get; }
+
+
+        public EndpointInstanceId(string endpointName, string instanceId) 
+            : this(endpointName, instanceId, instanceId)
+        {
+            
+        }
+
+        public EndpointInstanceId(string endpointName, string instanceId, string instanceName)
+        {
+            EndpointName = endpointName;
+            InstanceId = instanceId;
+            InstanceName = instanceName;
+        }
+
+        public static EndpointInstanceId From(IReadOnlyDictionary<string, string> headers)
+        {
+            var endpointName = headers[Headers.OriginatingEndpoint];
+            string instanceId;
+
+            if (headers.TryGetValue(MetricHeaders.MetricInstanceId, out instanceId))
+            {
+                return new EndpointInstanceId(endpointName, instanceId);    
+            }
+
+            var hostId = headers[Headers.OriginatingHostId];
+            var hostDisplayName = headers[Headers.HostDisplayName];
+
+            return new EndpointInstanceId(endpointName, hostId, hostDisplayName);
+        }
+
+        protected bool Equals(EndpointInstanceId other)
+        {
+            return string.Equals(EndpointName, other.EndpointName) && 
```

Comment:
I think it should be possible to host 2 endpoints in the same host. That can lead to a situation where two endpoints have different names but the same instance id.